From 784560b52083a029184f64a2c51e17fcf1bd1b59 Mon Sep 17 00:00:00 2001
From: William Wu <william.wu@rock-chips.com>
Date: Mon, 22 Jan 2018 15:09:35 +0800
Subject: [PATCH] usb: dwc2: don't clear mode if already in normal mode

When clear the force mode bits, it needs to wait up to
100 ms to account for any potential IDDIG filter delay.
In the current code, it will call this three times during
probe if dr_mode == OTG. With this patch, we can reduce
it to once. It can help to speed up the DWC2 probe time.

Change-Id: I588042b03b086599e4020c3072653dfb3138fdc1
Signed-off-by: William Wu <william.wu@rock-chips.com>
---
 drivers/usb/dwc2/core.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/usb/dwc2/core.c b/drivers/usb/dwc2/core.c
index d75bd5926e0b..a1fe052ccfc4 100644
--- a/drivers/usb/dwc2/core.c
+++ b/drivers/usb/dwc2/core.c
@@ -450,12 +450,15 @@ static void dwc2_clear_force_mode(struct dwc2_hsotg *hsotg)
 	u32 gusbcfg;
 
 	gusbcfg = dwc2_readl(hsotg->regs + GUSBCFG);
-	gusbcfg &= ~GUSBCFG_FORCEHOSTMODE;
-	gusbcfg &= ~GUSBCFG_FORCEDEVMODE;
-	dwc2_writel(gusbcfg, hsotg->regs + GUSBCFG);
 
-	if (dwc2_iddig_filter_enabled(hsotg))
-		msleep(100);
+	if (gusbcfg & (GUSBCFG_FORCEHOSTMODE | GUSBCFG_FORCEDEVMODE)) {
+		gusbcfg &= ~GUSBCFG_FORCEHOSTMODE;
+		gusbcfg &= ~GUSBCFG_FORCEDEVMODE;
+		dwc2_writel(gusbcfg, hsotg->regs + GUSBCFG);
+
+		if (dwc2_iddig_filter_enabled(hsotg))
+			msleep(100);
+	}
 }
 
 /*
-- 
2.35.3

